src/agent.py

```diff
@@ -53,6 +62,65 @@ def __init__(self, charm: ops.CharmBase, state: State):
         charm.framework.observe(
             charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
         )
+        # Event hooks for agent-discovery-ingress
+        charm.framework.observe(
+            ingress_observer.ingress.on.ready,
+            self.reconfigure_agent_discovery,
+        )
+        charm.framework.observe(
+            ingress_observer.ingress.on.revoked,
+            self.reconfigure_agent_discovery,
+        )
+
+    @property
+    def agent_discovery_url(self) -> str:
+        """Return the external hostname to be passed to agents via the integration.
+
+        If we do not have an ingress, then use the pod ip as hostname.
+        The reason to prefer this over the pod name (which is the actual
+        hostname visible from the pod) or a K8s service, is that those
+        are routable virtually exclusively inside the cluster (as they rely)
+        on the cluster's DNS service, while the ip address is _sometimes_
+        routable from the outside, e.g., when deploying on MicroK8s on Linux.
+
+        Returns:
+            The charm's agent discovery url.
+        """
+        # Check if an ingress URL is available
+        try:
+            if ingress_url := self.ingress_observer.ingress.url:
+                return ingress_url
+        except ops.ModelError as exc:
+            # We only log the error here as we can fallback to using pod IP
+            # if ingress is not available
+            logger.error(
+                "Failed obtaining agent discovery url: %s, is the charm shutting down?", exc
+            )
+
+        # Fallback to pod IP
+        if binding := self.charm.model.get_binding("juju-info"):
+            unit_ip = str(binding.network.bind_address)
+            if self._is_valid_ip_address(unit_ip):
+                return f"http://{unit_ip}:{jenkins.WEB_PORT}"
+
+        # Fallback to fqdn
+        return f"http://{socket.getfqdn()}:{jenkins.WEB_PORT}"
+
+    def _is_valid_ip_address(self, address: str) -> bool:
+        """Validate an IP address.
+
+        Args:
+            address: a string representing a unit address
+
+        Returns:
+            Whether the IP address is a valid unit IP address
+        """
+        try:
+            _ = ipaddress.ip_address(address)
```

Comment:
```suggestion
            ipaddress.ip_address(address)
```
I think the assignment is not necessary.